### ISSUE TITLE
Add a `--profile` option to `lxc image copy`

### ISSUE THREAD

**Issue by stgraber** [lxc/lxd#9236]
Images in LXD can have a set of profiles applied to them. By default it's just the `default` profile, but this can be changed through `lxc image edit`.

It would be nice if when copying an image with `lxc image copy`, the set of profiles could also be modified, something like:

```
lxc image copy images:ubuntu/20.04 local: --alias ubuntu --profile foo --profile bar
```

**Comment by ulziibuyan** [comment]
Hi, @stgraber. I'll grab this too. BTW, Did LXD opt-in for Hacktoberfest this year? Hoping my PR count towards the goal.

**Comment by stgraber** [comment]
We are indeed!

**Comment by stgraber** [comment]
@ulziibuyan are you still working on this one?

**Comment by andrewli6504** [comment]
Hello, I am a student at UT Austin taking Virtualization, and I see that there has not been work on this issue for a little while - Could me and @BlakeRomero I try tackling this issue?

**Comment by stgraber** [comment]
@andrewli6504 currently still assigned to @ulziibuyan, if he doesn't have the time anymore or is unresponsive, we can re-assign it to you.

**Comment by ulziibuyan** [comment]
Sorry for blocking this, I've not got the time I hoped. I've unassigned myself...

**Comment by stgraber** [comment]
No worries!

**Comment by andrewli6504** [comment]
As I was messing around with images and profiles to understand how they work, I found something that either is an issue, or I have misunderstood something. I created a new profile with `lxc profile create tempprofile`, then tried to add that profile to an image by editing the YML file by doing `lxc image edit ubuntu`, where ubuntu was the name of my image. Now, tempprofile shows up as a profile for the image, but when I do `lxc profile list` or `lxc profile show tempprofile`, they don't show that the profile is used by my image. Is there another way to apply a profile, or am I misunderstanding something else?

**Comment by stgraber** [comment]
Nope, what you did is perfectly correct, I think it's simply that we don't have UsedBy on profiles track images.
That's mostly because this is a softer relationship than an instance to a profile. That is, deleting a profile shouldn't be blocked because some images default to using it.

**Comment by andrewli6504** [comment]
Also, I am trying to figure out how to add a new command flag - where should I be looking to add that in?

**Comment by stgraber** [comment]
To add `--profile` to `lxc image copy`, you'd change `lxc/image.go`, look for the `cmdImageCopy` section and add a `flagProfile` section similar to what you'll find in `lxc/init.go` for example.

**Comment by andrewli6504** [comment]
I am trying to compile the code to see if the changes I implemented work by following the steps in installing.md. However, the last step, `sudo -E PATH=${PATH} LD_LIBRARY_PATH=${LD_LIBRARY_PATH} $(go env GOPATH)/bin/lxd --group sudo` I get the following errors:
`WARN[12-10|01:42:41]  - Couldn't find the CGroup blkio.weight, disk priority will be ignored 
WARN[12-10|01:42:41] Instance type not operational            err="KVM support is missing" type=virtual-machine driver=qemu
WARN[12-10|01:42:42] Firewall failed to detect any compatible driver, falling back to "xtables" (but some features may not work as expected due to: Backend command "ebtables" is an nftables shim) 
WARN[12-10|01:42:42] Failed to initialize fanotify, falling back on fsnotify err="Failed to watch directory \"/dev\": no such device"`
I am running on a VirtualBox VM. Previously, I was able to install LXD with apt install, but now trying to compile it myself it does not work.

**Comment by tomponline** [comment]
These are not errors they are warnings (hence the "WARN" prefix) and shouldn't prevent LXD from starting. What specifically doesn't work?

**Comment by andrewli6504** [comment]
When I try and run any commands it tells me that it cannot find the command lxc

**Comment by tomponline** [comment]
OK so thats not indicating that LXD cannot start, just that you're not actually running the `lxc` client command.

Where did you install `lxc` into, do you see it at `~/go/bin/lxc`?

**Comment by andrewli6504** [comment]
I do see it there

**Comment by tomponline** [comment]
Did you set these?

```
export PATH="${PATH}:$(go env GOPATH)/bin"
export LD_LIBRARY_PATH="$(go env GOPATH)/deps/dqlite/.libs/:$(go env GOPATH)/deps/raft/.libs/:${LD_LIBRARY_PATH}"
```

**Comment by andrewli6504** [comment]
I thought I did, but after doing it again it works. Sorry for the trouble.

**Comment by andrewli6504** [comment]
Right now, I can't get the copy command to take in a profile argument. I am currently doing this: `cmd.Flags().StringArrayVarP(&c.flagProfiles, "profile", "p", nil, i18n.G("Profile to apply to the new instance")+"``")` to get the profile from the user, and have added a flagProfile section as was suggested. Am I missing something?

**Comment by stgraber** [comment]
I don't know, it should be very straightforward to do, like here:
```
stgraber@lxd-build:~/go/src/github.com/lxc/lxd (stgraber/master)$ git diff
diff --git a/lxc/image.go b/lxc/image.go
index 6a033bb38..8cff132d0 100644
--- a/lxc/image.go
+++ b/lxc/image.go
@@ -134,6 +134,7 @@ type cmdImageCopy struct {
        flagAutoUpdate  bool
        flagVM          bool
        flagMode        string
+       flagProfile     []string
 }
 
 func (c *cmdImageCopy) Command() *cobra.Command {
@@ -153,6 +154,7 @@ It requires the source to be an alias and for it to be public.`))
        cmd.Flags().StringArrayVar(&c.flagAliases, "alias", nil, i18n.G("New aliases to add to the image")+"``")
        cmd.Flags().BoolVar(&c.flagVM, "vm", false, i18n.G("Copy virtual machine images"))
        cmd.Flags().StringVar(&c.flagMode, "mode", "pull", i18n.G("Transfer mode. One of pull (default), push or relay")+"``")
+       cmd.Flags().StringArrayVarP(&c.flagProfile, "profile", "p", nil, i18n.G("Profile to apply to the new instance")+"``")
        cmd.RunE = c.Run
 
        return cmd
@@ -161,6 +163,10 @@ It requires the source to be an alias and for it to be public.`))
 func (c *cmdImageCopy) Run(cmd *cobra.Command, args []string) error {
        conf := c.global.conf
 
+       if len(c.flagProfile) > 0 {
+               return fmt.Errorf("DEMO: profiles=%s", c.flagProfile)
+       }
+
        // Quick checks.
        exit, err := c.global.CheckArgs(cmd, args, 2, 2)
        if exit {
stgraber@lxd-build:~/go/src/github.com/lxc/lxd (stgraber/master)$ make
CC="cc" CGO_LDFLAGS_ALLOW="(-Wl,-wrap,pthread_create)|(-Wl,-z,now)" go install -v -tags "libsqlite3"  ./...
CGO_ENABLED=0 go install -v -tags netgo ./lxd-p2c
CGO_ENABLED=0 go install -v -tags agent,netgo ./lxd-agent
LXD built successfully
stgraber@lxd-build:~/go/src/github.com/lxc/lxd (stgraber/master)$ lxc image copy --profile foo --profile bar
Error: DEMO: profiles=[foo bar]
stgraber@lxd-build:~/go/src/github.com/lxc/lxd (stgraber/master)$ 
```

**Comment by stgraber** [comment]
@andrewli6504 @BlakeRomero hello, just wanted to check if you still intend to work on this or if we should unassign it so someone else can take a crack at it?

**Comment by andrewli6504** [comment]
Yes, I am still working on this issue.

**Comment by kamzar1** [comment]
Does a profile or profiles as image attribute matter at all?
Most you do with an image is creating an instance of it. In that moment, either 'default' profile is applied when not explicitly choosing a --profile during creation.
Should an image profile at some point override the chosen profile during creation, than it is relevant, otherwise not really.

**Comment by stgraber** [comment]
LXD has supported images having a different set of profiles for quite a while already, you can see it with `lxc image show` and change it with `lxc image edit`.

Those profiles assigned to an image only apply when the user doesn't provide their own set (`-p` during launch) and when not set will default to just the `default` profile.

**Comment by stgraber** [comment]
Cleaning assignment on this one as there's been no activity for a long time.

**Comment by presztak** [comment]
@tomponline Can I work on this?
